File: python_engine/workflow_engine/engine.py
"""
Core workflow engine logic.
"""

from __future__ import annotations

import json
import time
import uuid
from dataclasses import dataclass
from typing import Any, Callable, Dict, List, Optional

from .config import EngineConfig
from .logging_utils import get_logger
from .metrics import MetricsRegistry
from .models import HistoryEvent, RetryPolicy, Task
from .persistence import SqlitePersistence
from .queue import TaskQueue
from .retry import compute_backoff_seconds, should_retry
from .workflow import Command, DecisionResult, FailWorkflow, ScheduleActivity, ScheduleTimer, WorkflowContext, WorkflowDefinition


ActivityFunction = Callable[["ActivityContext", Dict[str, Any]], Dict[str, Any]]


@dataclass
class ActivityDefinition:
    name: str
    func: ActivityFunction
    retry_policy: RetryPolicy
    timeout_seconds: int
    heartbeat_timeout_seconds: int


class ActivityContext:
    def __init__(self, engine: "WorkflowEngine", task_id: str) -> None:
        self._engine = engine
        self._task_id = task_id

    def heartbeat(self) -> None:
        self._engine.queue.heartbeat(self._task_id)
        self._engine.metrics.counter("activity_heartbeats_total").inc()


class WorkflowEngine:
    def __init__(self, config: EngineConfig) -> None:
        self.config = config
        self.logger = get_logger("workflow_engine", level=config.log_level)
        self.persistence = SqlitePersistence(config.db_path)
        self.queue = TaskQueue(self.persistence)
        self.metrics = MetricsRegistry()
        self._workflows: Dict[str, WorkflowDefinition] = {}
        self._activities: Dict[str, ActivityDefinition] = {}

    def register_workflow(self, definition: WorkflowDefinition) -> None:
        self._workflows[definition.name] = definition

    def register_activity(
        self,
        name: str,
        func: ActivityFunction,
        retry_policy: Optional[RetryPolicy] = None,
        timeout_seconds: Optional[int] = None,
        heartbeat_timeout_seconds: Optional[int] = None,
    ) -> None:
        policy = retry_policy or RetryPolicy()
        self._activities[name] = ActivityDefinition(
            name=name,
            func=func,
            retry_policy=policy,
            timeout_seconds=timeout_seconds or self.config.workflow_task_timeout_seconds,
            heartbeat_timeout_seconds=heartbeat_timeout_seconds or self.config.activity_heartbeat_timeout_seconds,
        )

    def start_workflow(
        self,
        workflow_type: str,
        workflow_id: str,
        input_payload: Dict[str, Any],
        run_timeout_seconds: Optional[int] = None,
    ) -> str:
        if workflow_type not in self._workflows:
            raise ValueError(f"Unknown workflow type {workflow_type}")
        run_id = str(uuid.uuid4())
        definition = self._workflows[workflow_type]
        timeout = run_timeout_seconds or self.config.run_timeout_seconds
        self.persistence.create_run(
            workflow_id=workflow_id,
            workflow_type=workflow_type,
            run_id=run_id,
            state=definition.initial_state(),
            state_data={"sequence": 0},
            run_timeout_seconds=timeout,
            workflow_task_timeout_seconds=self.config.workflow_task_timeout_seconds,
            input_payload=input_payload,
        )
        self.enqueue_workflow_task(run_id)
        self.logger.info("Started workflow %s run %s", workflow_id, run_id)
        return run_id

    def enqueue_workflow_task(self, run_id: str) -> str:
        return self.queue.enqueue_task(
            run_id=run_id,
            task_type="workflow_task",
            queue_name=self.config.workflow_task_queue,
            payload={"run_id": run_id},
            timeout_seconds=self.config.workflow_task_timeout_seconds,
            max_attempts=self.config.max_task_attempts,
        )

    def enqueue_activity_task(self, run_id: str, activity_name: str, payload: Dict[str, Any], timeout_seconds: int, heartbeat_timeout_seconds: int, max_attempts: int) -> str:
        return self.queue.enqueue_task(
            run_id=run_id,
            task_type="activity_task",
            queue_name=self.config.activity_task_queue,
            payload=payload,
            timeout_seconds=timeout_seconds,
            heartbeat_timeout_seconds=heartbeat_timeout_seconds,
            max_attempts=max_attempts,
        )

    def enqueue_timer_task(self, run_id: str, payload: Dict[str, Any], fire_at: float) -> str:
        return self.queue.enqueue_task(
            run_id=run_id,
            task_type="timer_task",
            queue_name=self.config.timer_task_queue,
            payload=payload,
            scheduled_at=fire_at,
            timeout_seconds=None,
            max_attempts=self.config.max_task_attempts,
        )

    def process_workflow_task(self, task: Task) -> None:
        run_id = task.payload["run_id"]
        run = self.persistence.get_run(run_id)
        if not run or run["status"] != "RUNNING":
            self.queue.complete_task(task.task_id)
            return

        self.persistence.append_event(run_id, "WorkflowTaskStarted", {"task_id": task.task_id})
        history = self.persistence.list_events(run_id)
        definition = self._workflows[run["workflow_type"]]
        ctx = WorkflowContext(run_id, run["state_data"], history)
        decision = definition.decide(ctx, run["state"], ctx.state_data, history)

        self._apply_decision(run_id, decision, history)
        self.queue.complete_task(task.task_id)
        self.metrics.counter("workflow_tasks_completed_total").inc()

    def _apply_decision(self, run_id: str, decision: DecisionResult, history: List[HistoryEvent]) -> None:
        status = "RUNNING"
        terminal_from_command: Optional[str] = None
        for command in decision.commands:
            self._apply_command(run_id, command, history)
            if command.command_type == "CompleteWorkflow":
                terminal_from_command = "COMPLETED"
            if command.command_type == "FailWorkflow":
                terminal_from_command = "FAILED"

        if terminal_from_command:
            status = terminal_from_command
        elif decision.complete:
            status = "COMPLETED"
            self.persistence.append_event(run_id, "WorkflowCompleted", {"result": {"status": "completed"}})
        elif decision.failure:
            status = "FAILED"
            self.persistence.append_event(run_id, "WorkflowFailed", {"error": decision.failure})

        self.persistence.update_run_state(run_id, decision.next_state, decision.state_data, status=status)
        self.persistence.append_event(
            run_id,
            "WorkflowTaskCompleted",
            {"state": decision.next_state, "commands": [cmd.command_type for cmd in decision.commands]},
        )

    def _apply_command(self, run_id: str, command: Command, history: List[HistoryEvent]) -> None:
        if command.command_type == "ScheduleActivity":
            self._apply_schedule_activity(run_id, command, history)
        elif command.command_type == "ScheduleTimer":
            self._apply_schedule_timer(run_id, command, history)
        elif command.command_type == "CompleteWorkflow":
            self.persistence.append_event(run_id, "WorkflowCompleted", {"result": command.attributes["result"]})
        elif command.command_type == "FailWorkflow":
            self.persistence.append_event(run_id, "WorkflowFailed", {"error": command.attributes["error"]})

    def _history_has_command(self, history: List[HistoryEvent], event_type: str, command_id: str) -> bool:
        return any(
            event.event_type == event_type and event.attributes.get("command_id") == command_id
            for event in history
        )

    def _apply_schedule_activity(self, run_id: str, command: Command, history: List[HistoryEvent]) -> None:
        command_id = command.attributes.get("command_id")
        if not command_id:
            raise ValueError("ScheduleActivity requires command_id for determinism")
        if self._history_has_command(history, "ActivityScheduled", command_id):
            return

        activity_name = command.attributes["activity_name"]
        if activity_name not in self._activities:
            raise ValueError(f"Unknown activity {activity_name}")
        activity_id = command.attributes["activity_id"]
        input_payload = command.attributes["input"]
        retry_policy = RetryPolicy(**command.attributes["retry_policy"])
        timeout_seconds = int(command.attributes["timeout_seconds"])
        heartbeat_timeout_seconds = int(command.attributes["heartbeat_timeout_seconds"])

        self.persistence.append_event(
            run_id,
            "ActivityScheduled",
            {
                "command_id": command_id,
                "activity_id": activity_id,
                "activity_name": activity_name,
                "input": input_payload,
                "timeout_seconds": timeout_seconds,
                "heartbeat_timeout_seconds": heartbeat_timeout_seconds,
                "retry_policy": retry_policy.__dict__,
            },
        )
        self.enqueue_activity_task(
            run_id=run_id,
            activity_name=activity_name,
            payload={
                "run_id": run_id,
                "activity_id": activity_id,
                "activity_name": activity_name,
                "input": input_payload,
                "retry_policy": retry_policy.__dict__,
            },
            timeout_seconds=timeout_seconds,
            heartbeat_timeout_seconds=heartbeat_timeout_seconds,
            max_attempts=retry_policy.max_attempts,
        )

    def _apply_schedule_timer(self, run_id: str, command: Command, history: List[HistoryEvent]) -> None:
        command_id = command.attributes.get("command_id")
        if not command_id:
            raise ValueError("ScheduleTimer requires command_id for determinism")
        if self._history_has_command(history, "TimerScheduled", command_id):
            return
        timer_id = command.attributes["timer_id"]
        timeout_seconds = int(command.attributes["timeout_seconds"])
        fire_at = time.time() + timeout_seconds
        self.persistence.append_event(
            run_id,
            "TimerScheduled",
            {"command_id": command_id, "timer_id": timer_id, "fire_at": fire_at},
        )
        self.enqueue_timer_task(run_id, {"run_id": run_id, "timer_id": timer_id}, fire_at=fire_at)

    def record_activity_started(self, run_id: str, activity_id: str, task_id: str, attempt: int) -> None:
        self.persistence.append_event(
            run_id,
            "ActivityStarted",
            {"activity_id": activity_id, "task_id": task_id, "attempt": attempt},
        )

    def record_activity_completed(self, run_id: str, activity_id: str, result: Dict[str, Any], task_id: str, attempt: int) -> None:
        self.persistence.append_event(
            run_id,
            "ActivityCompleted",
            {"activity_id": activity_id, "result": result, "task_id": task_id, "attempt": attempt},
        )
        self.enqueue_workflow_task(run_id)

    def record_activity_failed(self, run_id: str, activity_id: str, error: str, task_id: str, attempt: int, retry_policy: RetryPolicy) -> None:
        self.persistence.append_event(
            run_id,
            "ActivityFailed",
            {"activity_id": activity_id, "error": error, "task_id": task_id, "attempt": attempt},
        )
        if should_retry(attempt, retry_policy):
            backoff = compute_backoff_seconds(attempt, retry_policy)
            scheduled_at = time.time() + backoff
            self.logger.warning("Retrying activity %s in %.2fs", activity_id, backoff)
            self.queue.reschedule_task(task_id, scheduled_at=scheduled_at)
            self.persistence.append_event(
                run_id,
                "ActivityRetryScheduled",
                {"activity_id": activity_id, "attempt": attempt + 1, "backoff_seconds": backoff},
            )
        else:
            self.enqueue_workflow_task(run_id)

    def record_timer_fired(self, run_id: str, timer_id: str) -> None:
        self.persistence.append_event(
            run_id,
            "TimerFired",
            {"timer_id": timer_id},
        )
        self.enqueue_workflow_task(run_id)

    def record_task_timeout(self, task_id: str) -> None:
        task = self.queue.get_task(task_id)
        if not task:
            return
        self.queue.fail_task(task_id, "timeout")
        if task.task_type == "activity_task":
            retry_policy = RetryPolicy(**task.payload.get("retry_policy", {}))
            self.record_activity_failed(
                run_id=task.run_id,
                activity_id=task.payload.get("activity_id", ""),
                error="timeout",
                task_id=task_id,
                attempt=task.attempt,
                retry_policy=retry_policy,
            )

    def record_heartbeat_timeout(self, task_id: str) -> None:
        task = self.queue.get_task(task_id)
        if not task:
            return
        self.queue.fail_task(task_id, "heartbeat_timeout")
        if task.task_type == "activity_task":
            retry_policy = RetryPolicy(**task.payload.get("retry_policy", {}))
            self.record_activity_failed(
                run_id=task.run_id,
                activity_id=task.payload.get("activity_id", ""),
                error="heartbeat_timeout",
                task_id=task_id,
                attempt=task.attempt,
                retry_policy=retry_policy,
            )

    def record_run_timeout(self, run_id: str) -> None:
        self.persistence.append_event(run_id, "WorkflowFailed", {"error": "run_timeout"})
        self.persistence.update_run_state(run_id, "FAILED", {}, status="FAILED")

File: go_engine/engine/engine.go
package engine

import (
	"fmt"
	"time"

	"github.com/google/uuid"
)

type ActivityFunc func(ctx *ActivityContext, input map[string]interface{}) (map[string]interface{}, error)

type ActivityDefinition struct {
	Name                    string
	Func                    ActivityFunc
	RetryPolicy             RetryPolicy
	TimeoutSeconds          int
	HeartbeatTimeoutSeconds int
}

type ActivityContext struct {
	engine *WorkflowEngine
	taskID string
}

func (c *ActivityContext) Heartbeat() {
	_ = c.engine.Queue.Heartbeat(c.taskID)
	c.engine.Metrics.Counter("activity_heartbeats_total").Inc("default", 1)
}

type WorkflowEngine struct {
	Config      Config
	Logger      *Logger
	Persistence *SqlitePersistence
	Queue       *TaskQueue
	Metrics     *MetricsRegistry
	workflows   map[string]WorkflowDefinition
	activities  map[string]ActivityDefinition
}

func NewWorkflowEngine(config Config) (*WorkflowEngine, error) {
	persistence, err := NewSqlitePersistence(config.DBPath)
	if err != nil {
		return nil, err
	}
	engine := &WorkflowEngine{
		Config:      config,
		Logger:      NewLogger("engine"),
		Persistence: persistence,
		Queue:       NewTaskQueue(persistence),
		Metrics:     NewMetricsRegistry(),
		workflows:   map[string]WorkflowDefinition{},
		activities:  map[string]ActivityDefinition{},
	}
	return engine, nil
}

func (e *WorkflowEngine) RegisterWorkflow(def WorkflowDefinition) {
	e.workflows[def.Name()] = def
}

func (e *WorkflowEngine) RegisterActivity(name string, fn ActivityFunc, retryPolicy *RetryPolicy, timeoutSeconds *int, heartbeatTimeoutSeconds *int) {
	policy := DefaultRetryPolicy()
	if retryPolicy != nil {
		policy = *retryPolicy
	}
	timeout := e.Config.WorkflowTaskTimeoutSeconds
	if timeoutSeconds != nil {
		timeout = *timeoutSeconds
	}
	heartbeat := e.Config.ActivityHeartbeatTimeoutSeconds
	if heartbeatTimeoutSeconds != nil {
		heartbeat = *heartbeatTimeoutSeconds
	}
	e.activities[name] = ActivityDefinition{
		Name:                    name,
		Func:                    fn,
		RetryPolicy:             policy,
		TimeoutSeconds:          timeout,
		HeartbeatTimeoutSeconds: heartbeat,
	}
}

func (e *WorkflowEngine) StartWorkflow(workflowType, workflowID string, inputPayload map[string]interface{}, runTimeoutSeconds *int) (string, error) {
	definition, ok := e.workflows[workflowType]
	if !ok {
		return "", fmt.Errorf("unknown workflow type %s", workflowType)
	}
	runID := uuid.NewString()
	timeout := e.Config.RunTimeoutSeconds
	if runTimeoutSeconds != nil {
		timeout = *runTimeoutSeconds
	}
	err := e.Persistence.CreateRun(workflowID, workflowType, runID, definition.InitialState(), map[string]interface{}{"sequence": 0}, timeout, e.Config.WorkflowTaskTimeoutSeconds, inputPayload)
	if err != nil {
		return "", err
	}
	_, err = e.EnqueueWorkflowTask(runID)
	if err != nil {
		return "", err
	}
	return runID, nil
}

func (e *WorkflowEngine) EnqueueWorkflowTask(runID string) (string, error) {
	timeout := e.Config.WorkflowTaskTimeoutSeconds
	return e.Queue.EnqueueTask(runID, "workflow_task", e.Config.WorkflowTaskQueue, map[string]interface{}{"run_id": runID}, nil, &timeout, nil, e.Config.MaxTaskAttempts)
}

func (e *WorkflowEngine) EnqueueActivityTask(runID string, payload map[string]interface{}, timeoutSeconds int, heartbeatTimeoutSeconds int, maxAttempts int) (string, error) {
	return e.Queue.EnqueueTask(runID, "activity_task", e.Config.ActivityTaskQueue, payload, nil, &timeoutSeconds, &heartbeatTimeoutSeconds, maxAttempts)
}

func (e *WorkflowEngine) EnqueueTimerTask(runID string, payload map[string]interface{}, fireAt float64) (string, error) {
	return e.Queue.EnqueueTask(runID, "timer_task", e.Config.TimerTaskQueue, payload, &fireAt, nil, nil, e.Config.MaxTaskAttempts)
}

func (e *WorkflowEngine) ProcessWorkflowTask(task Task) error {
	run, err := e.Persistence.GetRun(task.RunID)
	if err != nil {
		return err
	}
	if run["status"] != "RUNNING" {
		_ = e.Queue.CompleteTask(task.TaskID)
		return nil
	}
	_, _ = e.Persistence.AppendEvent(task.RunID, "WorkflowTaskStarted", map[string]interface{}{"task_id": task.TaskID})
	history, err := e.Persistence.ListEvents(task.RunID)
	if err != nil {
		return err
	}
	def := e.workflows[run["workflow_type"].(string)]
	stateData := run["state_data"].(map[string]interface{})
	ctx := NewWorkflowContext(task.RunID, stateData, history)
	decision := def.Decide(ctx, run["state"].(string), ctx.StateData, history)
	e.applyDecision(task.RunID, decision, history)
	_ = e.Queue.CompleteTask(task.TaskID)
	e.Metrics.Counter("workflow_tasks_completed_total").Inc("default", 1)
	return nil
}

func (e *WorkflowEngine) applyDecision(runID string, decision DecisionResult, history []HistoryEvent) {
	status := "RUNNING"
	terminalFromCommand := ""
	for _, cmd := range decision.Commands {
		e.applyCommand(runID, cmd, history)
		if cmd.CommandType == "CompleteWorkflow" {
			terminalFromCommand = "COMPLETED"
		}
		if cmd.CommandType == "FailWorkflow" {
			terminalFromCommand = "FAILED"
		}
	}
	if terminalFromCommand != "" {
		status = terminalFromCommand
	} else if decision.Complete {
		status = "COMPLETED"
		_, _ = e.Persistence.AppendEvent(runID, "WorkflowCompleted", map[string]interface{}{"result": map[string]interface{}{"status": "completed"}})
	} else if decision.Failure != "" {
		status = "FAILED"
		_, _ = e.Persistence.AppendEvent(runID, "WorkflowFailed", map[string]interface{}{"error": decision.Failure})
	}
	_ = e.Persistence.UpdateRunState(runID, decision.NextState, decision.StateData, status)
	_, _ = e.Persistence.AppendEvent(runID, "WorkflowTaskCompleted", map[string]interface{}{
		"state":    decision.NextState,
		"commands": e.commandTypes(decision.Commands),
	})
}

func (e *WorkflowEngine) commandTypes(commands []Command) []string {
	out := make([]string, 0, len(commands))
	for _, cmd := range commands {
		out = append(out, cmd.CommandType)
	}
	return out
}

func (e *WorkflowEngine) applyCommand(runID string, command Command, history []HistoryEvent) {
	switch command.CommandType {
	case "ScheduleActivity":
		e.applyScheduleActivity(runID, command, history)
	case "ScheduleTimer":
		e.applyScheduleTimer(runID, command, history)
	case "CompleteWorkflow":
		_, _ = e.Persistence.AppendEvent(runID, "WorkflowCompleted", map[string]interface{}{"result": command.Attributes["result"]})
	case "FailWorkflow":
		_, _ = e.Persistence.AppendEvent(runID, "WorkflowFailed", map[string]interface{}{"error": command.Attributes["error"]})
	}
}

func (e *WorkflowEngine) historyHasCommand(history []HistoryEvent, eventType, commandID string) bool {
	for _, event := range history {
		if event.EventType == eventType {
			if cmdID, ok := event.Attributes["command_id"].(string); ok && cmdID == commandID {
				return true
			}
		}
	}
	return false
}

func (e *WorkflowEngine) applyScheduleActivity(runID string, command Command, history []HistoryEvent) {
	commandID, ok := command.Attributes["command_id"].(string)
	if !ok || commandID == "" {
		panic("ScheduleActivity requires command_id")
	}
	if e.historyHasCommand(history, "ActivityScheduled", commandID) {
		return
	}
	activityName, _ := command.Attributes["activity_name"].(string)
	if _, ok := e.activities[activityName]; !ok {
		panic(fmt.Sprintf("unknown activity %s", activityName))
	}
	activityID, _ := command.Attributes["activity_id"].(string)
	input, _ := command.Attributes["input"].(map[string]interface{})
	timeoutSeconds := intFromAttribute(command.Attributes["timeout_seconds"])
	heartbeatSeconds := intFromAttribute(command.Attributes["heartbeat_timeout_seconds"])
	retryPolicy := retryPolicyFromAttribute(command.Attributes["retry_policy"])

	_, _ = e.Persistence.AppendEvent(runID, "ActivityScheduled", map[string]interface{}{
		"command_id":               commandID,
		"activity_id":              activityID,
		"activity_name":            activityName,
		"input":                    input,
		"timeout_seconds":          timeoutSeconds,
		"heartbeat_timeout_seconds": heartbeatSeconds,
		"retry_policy":             retryPolicy,
	})
	_, _ = e.EnqueueActivityTask(runID, map[string]interface{}{
		"run_id":        runID,
		"activity_id":   activityID,
		"activity_name": activityName,
		"input":         input,
		"retry_policy":  retryPolicy,
	}, timeoutSeconds, heartbeatSeconds, retryPolicy.MaxAttempts)
}

func (e *WorkflowEngine) applyScheduleTimer(runID string, command Command, history []HistoryEvent) {
	commandID, ok := command.Attributes["command_id"].(string)
	if !ok || commandID == "" {
		panic("ScheduleTimer requires command_id")
	}
	if e.historyHasCommand(history, "TimerScheduled", commandID) {
		return
	}
	timerID, _ := command.Attributes["timer_id"].(string)
	timeoutSeconds := intFromAttribute(command.Attributes["timeout_seconds"])
	fireAt := float64(time.Now().UnixNano())/1e9 + float64(timeoutSeconds)
	_, _ = e.Persistence.AppendEvent(runID, "TimerScheduled", map[string]interface{}{
		"command_id": commandID,
		"timer_id":   timerID,
		"fire_at":    fireAt,
	})
	_, _ = e.EnqueueTimerTask(runID, map[string]interface{}{"run_id": runID, "timer_id": timerID}, fireAt)
}

func (e *WorkflowEngine) RecordActivityStarted(runID, activityID, taskID string, attempt int) {
	_, _ = e.Persistence.AppendEvent(runID, "ActivityStarted", map[string]interface{}{
		"activity_id": activityID,
		"task_id":     taskID,
		"attempt":     attempt,
	})
}

func (e *WorkflowEngine) RecordActivityCompleted(runID, activityID, taskID string, attempt int, result map[string]interface{}) {
	_, _ = e.Persistence.AppendEvent(runID, "ActivityCompleted", map[string]interface{}{
		"activity_id": activityID,
		"task_id":     taskID,
		"attempt":     attempt,
		"result":      result,
	})
	_, _ = e.EnqueueWorkflowTask(runID)
}

func (e *WorkflowEngine) RecordActivityFailed(runID, activityID, taskID string, attempt int, err string, retryPolicy RetryPolicy) {
	_, _ = e.Persistence.AppendEvent(runID, "ActivityFailed", map[string]interface{}{
		"activity_id": activityID,
		"task_id":     taskID,
		"attempt":     attempt,
		"error":       err,
	})
	if ShouldRetry(attempt, retryPolicy) {
		backoff := ComputeBackoffSeconds(attempt, retryPolicy)
		scheduledAt := float64(time.Now().UnixNano())/1e9 + backoff
		_ = e.Queue.RescheduleTask(taskID, scheduledAt)
		_, _ = e.Persistence.AppendEvent(runID, "ActivityRetryScheduled", map[string]interface{}{
			"activity_id":    activityID,
			"attempt":        attempt + 1,
			"backoff_seconds": backoff,
		})
	} else {
		_, _ = e.EnqueueWorkflowTask(runID)
	}
}

func (e *WorkflowEngine) RecordTimerFired(runID, timerID string) {
	_, _ = e.Persistence.AppendEvent(runID, "TimerFired", map[string]interface{}{
		"timer_id": timerID,
	})
	_, _ = e.EnqueueWorkflowTask(runID)
}

func (e *WorkflowEngine) RecordTaskTimeout(taskID string) {
	task, err := e.Queue.GetTask(taskID)
	if err != nil || task == nil {
		return
	}
	_ = e.Queue.FailTask(taskID, "timeout")
	if task.TaskType == "activity_task" {
		retryPolicy := retryPolicyFromAttribute(task.Payload["retry_policy"])
		activityID, _ := task.Payload["activity_id"].(string)
		e.RecordActivityFailed(task.RunID, activityID, taskID, task.Attempt, "timeout", retryPolicy)
	}
}

func (e *WorkflowEngine) RecordHeartbeatTimeout(taskID string) {
	task, err := e.Queue.GetTask(taskID)
	if err != nil || task == nil {
		return
	}
	_ = e.Queue.FailTask(taskID, "heartbeat_timeout")
	if task.TaskType == "activity_task" {
		retryPolicy := retryPolicyFromAttribute(task.Payload["retry_policy"])
		activityID, _ := task.Payload["activity_id"].(string)
		e.RecordActivityFailed(task.RunID, activityID, taskID, task.Attempt, "heartbeat_timeout", retryPolicy)
	}
}

func (e *WorkflowEngine) RecordRunTimeout(runID string) {
	_, _ = e.Persistence.AppendEvent(runID, "WorkflowFailed", map[string]interface{}{"error": "run_timeout"})
	_ = e.Persistence.UpdateRunState(runID, "FAILED", map[string]interface{}{}, "FAILED")
}

func retryPolicyFromAttribute(value interface{}) RetryPolicy {
	switch v := value.(type) {
	case RetryPolicy:
		return v
	case map[string]interface{}:
		policy := DefaultRetryPolicy()
		if val, ok := v["initial_interval_seconds"].(float64); ok {
			policy.InitialIntervalSeconds = val
		}
		if val, ok := v["backoff_coefficient"].(float64); ok {
			policy.BackoffCoefficient = val
		}
		if val, ok := v["max_interval_seconds"].(float64); ok {
			policy.MaxIntervalSeconds = val
		}
		if val, ok := v["max_attempts"].(float64); ok {
			policy.MaxAttempts = int(val)
		}
		return policy
	default:
		return DefaultRetryPolicy()
	}
}

func intFromAttribute(value interface{}) int {
	switch v := value.(type) {
	case int:
		return v
	case float64:
		return int(v)
	default:
		return 0
	}
}
